crypto/rand/drbg_hash.c

```diff
@@ -0,0 +1,292 @@
+/*
+ * Copyright 2011-2018 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the OpenSSL license (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <stdlib.h>
+#include <string.h>
+#include <openssl/crypto.h>
+#include <openssl/err.h>
+#include <openssl/rand.h>
+#include "internal/thread_once.h"
+#include "rand_lcl.h"
+
+/* 440 bits from SP800-90Ar1 10.1 table 2 */
+#define HASH_PRNG_SMALL_SEEDLEN   (440/8)
+/* Determine what seedlen to use based on the block length */
+#define MAX_BLOCKLEN_USING_SMALL_SEEDLEN (256/8)
+#define INBYTE_IGNORE ((unsigned char)0xFF)
+
+
+/* SP800-90Ar1 10.3.1 Derivation function using a Hash Function (Hash_df).
+ * The input string used is composed of:
+ *    inbyte - An optional leading byte (ignore if equal to INBYTE_IGNORE)
+ *    in - input string 1 (A Non NULL value).
+ *    in2 - optional input string (Can be NULL).
+ *    in3 - optional input string (Can be NULL).
+ *    These are concatenated as part of the DigestUpdate process.
+ */
+static int hash_df(RAND_DRBG *drbg, unsigned char *out,
+                   const unsigned char inbyte,
+                   const unsigned char *in, size_t inlen,
+                   const unsigned char *in2, size_t in2len,
+                   const unsigned char *in3, size_t in3len)
+{
+    RAND_DRBG_HASH *hash = &drbg->data.hash;
+    EVP_MD_CTX *ctx = hash->ctx;
+    unsigned char *vtmp = hash->vtmp;
+    unsigned char tmp[1+4+1];
+    unsigned char tmp_sz = 5;
+    /* Standard only ever needs seedlen bytes which is always less than
+     * maximum permitted so no need to check length.
+     */
+    size_t outlen = drbg->seedlen;
+    size_t num_bits_returned = outlen * 8;
+    /* (Step 3) counter=1 tmp[0] is the 8 bit counter. */
+    tmp[0] = 1;
+    /* tmp[1..4] is the fixed 32 bit no_of_bits_to_return */
+    tmp[1] = (num_bits_returned >> 24) & 0xff;
+    tmp[2] = (num_bits_returned >> 16) & 0xff;
+    tmp[3] = (num_bits_returned >> 8) & 0xff;
+    tmp[4] = num_bits_returned & 0xff;
+    /* Tack the additional input byte onto the end of tmp if it exists */
+    if (inbyte != INBYTE_IGNORE)
+        tmp[tmp_sz++] = inbyte;
+
+    /* (Step 4) */
+    for (;;) {
+        /* (Step 4.1) temp = temp || Hash(counter||num_bits_returned||input */
+        if (!(EVP_DigestInit_ex(ctx, hash->md, NULL)
+                && EVP_DigestUpdate(ctx, tmp, tmp_sz)
+                && EVP_DigestUpdate(ctx, in, inlen)
+                && (in2 == NULL || EVP_DigestUpdate(ctx, in2, in2len))
+                && (in3 == NULL || EVP_DigestUpdate(ctx, in3, in3len))))
+            return 0;
```

Comment:
It's surprising we don't add an error to the queue here. But the ctr drbg doesn't seem to either.